File: server/protocols/marshaller.py
from __future__ import annotations
import struct

class BufferWriter:
    """
    Serializes Python values into bytes using network byte order (big-endian).
    Suitable for UDP request/reply payload construction.
    """
    def __init__(self) -> None:
        self._buffer = bytearray()

    def write_u8(self, value: int) -> None:
        if not 0 <= value <= 0xFF:
            raise ValueError(f"u8 out of range: {value}")
        self._buffer += struct.pack("!B", value)

    def write_u32(self, value: int) -> None:
        if not 0 <= value <= 0xFFFFFFFF:
            raise ValueError(f"u32 out of range: {value}")
        self._buffer += struct.pack("!I", value)

    def write_f64(self, value: float) -> None:
        self._buffer += struct.pack("!d", float(value))

    def write_bytes(self, value: bytes) -> None:
        self.write_u32(len(value))
        self._buffer += value

    def write_str(self, value: str) -> None:
        encoded = value.encode("utf-8")
        self.write_u32(len(encoded))
        self._buffer += encoded

    def to_bytes(self) -> bytes:
        return bytes(self._buffer)

class BufferReader:
    """
    Deserializes bytes into Python values using network byte order (big-endian).
    """
    def __init__(self, data: bytes) -> None:
        self._buffer = data
        self._offset = 0

    def _read_exact(self, size: int) -> bytes:
        end = self._offset + size
        if end > len(self._buffer):
            raise ValueError(
                f"Not enough bytes to read {size} byte(s). "
                f"Offset={self._offset}, buffer_length={len(self._buffer)}"
            )
        chunk = self._buffer[self._offset:end]
        self._offset = end
        return chunk

    def read_u8(self) -> int:
        return struct.unpack("!B", self._read_exact(1))[0]

    def read_u32(self) -> int:
        return struct.unpack("!I", self._read_exact(4))[0]

    def read_f64(self) -> float:
        return struct.unpack("!d", self._read_exact(8))[0]

    def read_bytes(self) -> bytes:
        length = self.read_u32()
        return self._read_exact(length)

    def read_str(self) -> str:
        length = self.read_u32()
        raw = self._read_exact(length)
        return raw.decode("utf-8")

    def remaining(self) -> int:
        return len(self._buffer) - self._offset

    def has_remaining(self) -> bool:
        return self.remaining() > 0

File: client/protocols/marshaller.js
module.exports = {
    BufferReader: class BufferReader {
        constructor(data) {
            this.data = data;
            this.offset = 0;
        }

        read(size) {
            const end = this.offset + size;
            if (end > this.data.length) {
                throw new Error(`Not enough bytes to read ${size} byte(s).
                    Offset=${this.offset}, buffer_length=${this.data.length}`);
            }
            const chunk = this.data.slice(this.offset, end);
            this.offset = end;
            return chunk;
        }

        readU8() {
            const value = this.read(1).readUInt8(0);
            return value;
        }

        readU32() {
            const value = this.read(4).readUInt32BE(0);
            return value;
        }

        readF64() {
            const value = this.read(8).readDoubleBE(0);
            return value;
        }

        readString() {
            const length = this.readU32();
            const strBuf = this.read(length);
            return strBuf.toString('utf8');
        }

        readBytes() {
            const length = this.readU32();
            return this.read(length);
        }

        remaining() {
            return this.data.length - this.offset;
        }

        hasRemaining() {
            return this.remaining() > 0;
        }
    },
    BufferWriter: class BufferWriter {
        constructor() {
            this.parts = [];
        }

        writeU8(value) {
            const buf = Buffer.alloc(1);
            buf.writeUInt8(value, 0);
            this.parts.push(buf);
        }

        writeU32(value) {
            const buf = Buffer.alloc(4);
            buf.writeUInt32BE(value, 0);
            this.parts.push(buf);
        }

        writeF64(value) {
            const buf = Buffer.alloc(8);
            buf.writeDoubleBE(value, 0);
            this.parts.push(buf);
        }

        writeString(value) {
            const strBuf = Buffer.from(value, 'utf8');
            this.writeU32(strBuf.length);
            this.parts.push(strBuf);
        }

        writeBytes(value) {
            this.writeU32(value.length);
            this.parts.push(value);
        }

        toBuffer() {
            return Buffer.concat(this.parts);
        }
    }
}